src/librustc_lint/types.rs

```diff
@@ -832,9 +833,22 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
             },
 
             ty::RawPtr(ty::TypeAndMut { ty, .. }) |
-            ty::Ref(_, ty, _) => self.check_type_for_ffi(cache, ty),
+            ty::Ref(_, ty, _) => self.check_type_for_ffi(cache, ty, false),
 
-            ty::Array(ty, _) => self.check_type_for_ffi(cache, ty),
+            ty::Array(inner_ty, _) => {
+                // C doesn't really support passing arrays by value.
+                // The only way to pass an array by value is through a struct.
+                if top_level {
```

Comment:
This is probably more readable as a guard:
````rust
ty::Array(..) if top_level => { .. }
ty::Array(inner_ty, _) => { .. }
```

Reply:
Done